snapcraft/internal/project_loader/_config.py

```diff
@@ -71,6 +72,106 @@ def _validate_epoch(instance):
     return True
 
 
+@jsonschema.FormatChecker.cls_checks('architectures')
+def _validate_architectures(instance):
+    standalone_build_ons = collections.Counter()
+    build_ons = collections.Counter()
+    run_ons = collections.Counter()
+
+    saw_strings = False
+    saw_dicts = False
+
+    for mapping in instance:
+        # This could either be a dict or a string. In the latter case, the
+        # schema will take care of it. We just need to further validate the
+        # dict.
+        if isinstance(mapping, str):
+            saw_strings = True
+        elif isinstance(mapping, dict):
+            saw_dicts = True
+            build_on = _get_architectures_list(mapping, 'build-on')
+            build_ons.update(build_on)
+
+            # Add to the list of run-ons. However, if no run-on is specified,
+            # we know it's implicitly the value of build-on, so use that
+            # for validation instead.
+            run_on = _get_architectures_list(mapping, 'run-on')
+            if run_on:
+                run_ons.update(run_on)
+            else:
+                standalone_build_ons.update(build_on)
+
+    # Architectures can either be a list of strings, or a list of objects.
+    # Mixing the two forms is unsupported.
+    if saw_strings and saw_dicts:
+        raise jsonschema.exceptions.ValidationError(
+            'every item must either be a string or an object',
+            path=['architectures'], instance=mapping)
+
+    # At this point, individual build-ons and run-ons have been validated,
+    # we just need to validate them across each other.
+
+    # First of all, if we have a `run-on: [all]` (or a standalone
+    # `build-on: [all]`) then we should only have one item in the instance,
+    # otherwise we know we'll have multiple snaps claiming they run on the same
+    # architectures (i.e. all and something else).
+    number_of_snaps = len(instance)
+    if 'all' in run_ons and number_of_snaps > 1:
+        raise jsonschema.exceptions.ValidationError(
+            "one of the items has 'all' in 'run-on', which only makes "
+            "sense if it's the only item, and there are {}".format(
+                number_of_snaps),
+            path=['architectures'], instance=instance)
+    if 'all' in build_ons and number_of_snaps > 1:
+        raise jsonschema.exceptions.ValidationError(
+            "one of the items has 'all' in 'build-on', which only makes "
+            "sense if it's the only item, and there are {}".format(
+                number_of_snaps),
+            path=['architectures'], instance=instance)
+
+    # We want to ensure that multiple `run-on`s (or standalone `build-on`s)
+    # don't incude the same arch, or they'll clash with each other when
+    # releasing.
+    all_ons = run_ons + standalone_build_ons
```

Comment:
this is really `all_run_ons` given that you are checking where to run. I think it would be easier to parse the intent with that name. Or maybe `where_to_run` or something involving a run :-)

Reply:
Indeed, fixed.